File: lib/services/executors/claude-agent/src/error-classify.ts
// Copyright © 2026 Fall Guy Consulting.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.apache at the repo root.

export type ClassifiedAgentError = { errorClass: string } | null;

export function classifyAgentError(
  stderr: string,
  exitCode: number | null,
): ClassifiedAgentError {
  void exitCode;
  if (!stderr) return null;
  const lower = stderr.toLowerCase();

  if (
    lower.includes("context_length_exceeded") ||
    lower.includes("context window") ||
    lower.includes("context_window") ||
    lower.includes("maximum context") ||
    lower.includes("prompt is too long")
  ) {
    return { errorClass: "agent/context_exceeded" };
  }

  if (lower.includes("tool_use_failed") || lower.includes("tool execution failed")) {
    const tool = parseToolName(stderr) ?? "unknown";
    return { errorClass: `agent/tool_use_failed/${tool}` };
  }

  if (
    lower.includes("(refusal)") ||
    lower.includes("refused by the model") ||
    lower.includes("declined to respond") ||
    /\brefusal\b/.test(lower)
  ) {
    return { errorClass: "agent/refused" };
  }

  return null;
}

function parseToolName(stderr: string): string | null {
  const quoted = /tool\s+["'`]([^"'`]+)["'`]/i.exec(stderr);
  if (quoted && quoted[1]!.length > 0) return quoted[1]!;

  const colon = /tool_use_failed[:\s]+["'`]?([A-Za-z0-9_./-]+)["'`]?/i.exec(stderr);
  if (colon && colon[1]!.length > 0 && colon[1]!.toLowerCase() !== "tool") {
    return colon[1]!;
  }

  const bare = /\btool\s+([A-Za-z0-9_./-]+)\s+(?:returned|failed|errored)/i.exec(stderr);
  if (bare && bare[1]!.length > 0) return bare[1]!;

  return null;
}

File: lib/services/executors/claude-agent/errorclassify.go
// Copyright © 2026 Fall Guy Consulting.
// SPDX-License-Identifier: AGPL-3.0-or-later OR LicenseRef-FallGuy-Commercial

package claudeagent

import (
	"regexp"
	"strings"
)

func ClassifyAgentError(stderr string) string {
	if stderr == "" {
		return ""
	}
	lower := strings.ToLower(stderr)

	if strings.Contains(lower, "context_length_exceeded") ||
		strings.Contains(lower, "context window") ||
		strings.Contains(lower, "context_window") ||
		strings.Contains(lower, "maximum context") ||
		strings.Contains(lower, "prompt is too long") {
		return "agent/context_exceeded"
	}

	if strings.Contains(lower, "tool_use_failed") || strings.Contains(lower, "tool execution failed") {
		tool := parseToolName(stderr)
		if tool == "" {
			tool = "unknown"
		}
		return "agent/tool_use_failed/" + tool
	}

	if strings.Contains(lower, "(refusal)") ||
		strings.Contains(lower, "refused by the model") ||
		strings.Contains(lower, "declined to respond") ||
		(refusalWordRe.MatchString(lower) && !refusalNegatedRe.MatchString(lower)) {
		return "agent/refused"
	}

	return ""
}

var (
	refusalWordRe    = regexp.MustCompile(`\brefusal\b`)
	refusalNegatedRe = regexp.MustCompile(`(?i)\b(?:no|not|non|isn'?t|wasn'?t|without)\b(?:\s+\S+){0,3}\s+refusal\b`)
	toolQuotedRe     = regexp.MustCompile("(?i)tool\\s+[\"'`]([^\"'`]+)[\"'`]")
	toolColonRe      = regexp.MustCompile("(?i)tool_use_failed[:\\s]+[\"'`]?([A-Za-z0-9_./-]+)[\"'`]?")
	toolBareRe       = regexp.MustCompile(`(?i)\btool\s+([A-Za-z0-9_./-]+)\s+(?:returned|failed|errored)`)
)

func parseToolName(stderr string) string {
	if m := toolQuotedRe.FindStringSubmatch(stderr); m != nil && m[1] != "" {
		return m[1]
	}
	if m := toolColonRe.FindStringSubmatch(stderr); m != nil && m[1] != "" && strings.ToLower(m[1]) != "tool" {
		return m[1]
	}
	if m := toolBareRe.FindStringSubmatch(stderr); m != nil && m[1] != "" {
		return m[1]
	}
	return ""
}
